### ISSUE TITLE
Make timeseries_dataset_from_array() more intuitive

### ISSUE THREAD

**Issue by bibhas2** [tensorflow/tensorflow#44592]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.3.0
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
``timeseries_dataset_from_array()`` was introduced to replace ``tf.keras.preprocessing.sequence.TimeseriesGenerator``. But it is not a drop in replacement. In fact, as I will show below, it takes a bit of work to make ``timeseries_dataset_from_array()`` behave like ``TimeseriesGenerator``. Issue #40343 was created regarding this. But the resolution and suggested work around are problematic. This is why I am opening this issue.

Let's explain the issues with an example.

```python
features = np.arange(1, 10)
labels = features * -1

print("Features:", features)
print("Labels:", labels)
```

This prints:

```
Features: [1 2 3 4 5 6 7 8 9]
Labels: [-1 -2 -3 -4 -5 -6 -7 -8 -9]
```

We can use ``TimeseriesGenerator`` as follows.

```python
sequence_length=3

gen = tf.keras.preprocessing.sequence.TimeseriesGenerator(features, labels, sequence_length, batch_size=1)

for inputs, targets in gen:
  print("Input:", inputs, "Target:", targets)
```

This correctly prints:

```
Input: [[1 2 3]] Target: [-4]
Input: [[2 3 4]] Target: [-5]
Input: [[3 4 5]] Target: [-6]
Input: [[4 5 6]] Target: [-7]
Input: [[5 6 7]] Target: [-8]
Input: [[6 7 8]] Target: [-9]
```

Now, let's use ``timeseries_dataset_from_array()`` as a drop in replacement.

```python
ds = tf.keras.preprocessing.timeseries_dataset_from_array(features, labels, sequence_length, batch_size=1)

for inputs, targets in ds:
  print("Input:", inputs.numpy(), "Target:", targets.numpy())
```

This prints:

```
Input: [[1 2 3]] Target: [-1]
Input: [[2 3 4]] Target: [-2]
Input: [[3 4 5]] Target: [-3]
Input: [[4 5 6]] Target: [-4]
Input: [[5 6 7]] Target: [-5]
Input: [[6 7 8]] Target: [-6]
Input: [[7 8 9]] Target: [-7]
```

This is not expected. To be honest, I am not quite sure in what scenarios one would want the above behavior.

The [documentation](https://www.tensorflow.org/api_docs/python/tf/keras/preprocessing/timeseries_dataset_from_array) shows a way around this. We can follow the "Example 2" in the documentation and re-write our test like this:

```python
ds = tf.keras.preprocessing.timeseries_dataset_from_array(features[:-sequence_length], labels[sequence_length:], sequence_length, batch_size=1)

for inputs, targets in ds:
  print("Input:", inputs.numpy(), "Target:", targets.numpy())
```

This prints:

```
Input: [[1 2 3]] Target: [-4]
Input: [[2 3 4]] Target: [-5]
Input: [[3 4 5]] Target: [-6]
Input: [[4 5 6]] Target: [-7]
```

This is good. But we are now missing the last two sequences:

```
Input: [[5 6 7]] Target: [-8]
Input: [[6 7 8]] Target: [-9]
```

Issue #40343 suggests the use of ``np.roll()`` as a work around. The resolution comment says that ``np.roll()`` was added to the documentation. But I fail to see it there. Also the way ``np.roll()`` is used in that issue seems problematic. Below we re-write our test using the way I think ``np.roll()`` should be used.

```python
ds = tf.keras.preprocessing.timeseries_dataset_from_array(
  features[:-1], np.roll(labels, -sequence_length)[:-1], sequence_length, batch_size=1)

for inputs, targets in ds:
  print("Input:", inputs.numpy(), "Target:", targets.numpy())
```

This prints:

```
Input: [[1 2 3]] Target: [-4]
Input: [[2 3 4]] Target: [-5]
Input: [[3 4 5]] Target: [-6]
Input: [[4 5 6]] Target: [-7]
Input: [[5 6 7]] Target: [-8]
Input: [[6 7 8]] Target: [-9]
```

This is now functionally equivalent to ``TimeseriesGenerator``.

**Will this change the current api? How?**
The issue can be addressed either by enhancing the documentation or by enhancing the API. I will prefer the later. Perhaps a flag can be supplied to ``timeseries_dataset_from_array()`` that will make it behave like ``TimeseriesGenerator``.

**Who will benefit with this feature?**
Currently ``timeseries_dataset_from_array()`` requires a bit of work to make it work for what I will consider as most common time series problems. This makes the API unintuitive. At minimum the documentation needs to be enhanced. It will be better if the API can be changed so one can easily use it in most common scenarios.

**Any Other info.**

**Comment by ravikyram** [comment]
@bibhas2 

Please feel free to submit a PR if you have use cases that supports that feature.Thanks!

**Comment by bibhas2** [comment]
I had a chance to think more about this and look at several projects on the Internet. I think what we need is a higher level API that is quick to use in most time series problems. 

This [excellent tutorial](https://www.tensorflow.org/tutorials/structured_data/time_series) from the TF documentation had to write good bit of code around ``timeseries_dataset_from_array`` to make it work. At that point I wonder if we are getting much value out of ``timeseries_dataset_from_array`` or not. 

We are also seeing projects like [this](https://github.com/anencore94/SlidingWindowGenerator/blob/master/slidingwindow_generator/slidingwindow_generator.py) coming up to put a friendly face around the API. As you can see, the work there is non-trivial.

I will be inclined to keep using ``tf.keras.preprocessing.sequence.TimeseriesGenerator`` as long as it is not fully deprecated. But, sadly it does not support multi-step label output. An [issue](https://github.com/keras-team/governance/pull/18) was opened to address this gap. But the issue has been closed suggesting the use of ``timeseries_dataset_from_array`` as an alternative.

**Comment by fchollet** [comment]
Hi mate,

Thanks for the comments. Do remark that there's a timeseries regression example right there in the docstring:

```python
input_data = data[:-10]
targets = data[10:]
dataset = tf.keras.preprocessing.timeseries_dataset_from_array(
    input_data, targets, sequence_length=10)
for batch in dataset:
  inputs, targets = batch
  assert np.array_equal(inputs[0], data[:10])  # First sequence: steps [0-9]
  assert np.array_equal(targets[0], data[10])  # Corresponding target: step 10
  break
```

Now, you seem to have found this workflow on your own, and you note:

> But we are now missing the last two sequences

You're right, that's weird. There's actually no need on our part to force the input data and targets to have the same length. We can just use `min(num_seqs, len(targets))` where `num_seqs` is the number of sub sequences that can be generated from the data.

So I'll just go ahead and do that.

**Comment by fchollet** [comment]
Done

**Comment by summa-code** [comment]
I am surprised they closed, If they purpose of this API is to make sliding window, then why do they have a whole nine yard of data type checking and normalization calls in between ? This is make it not usable when we use with feature classification.

**Comment by bibhas2** [comment]
I see that the issue is now closed. It's not clear from the comment from @fchollet what the proposed resolution is.

Anyone serious about time series data should avoid ``timeseries_dataset_from_array()``. For two main reasons:

- The API needs a lot of work arounds (as explained in my original issue description).
- It works with in-memory Numpy array and hence not well suited for very large amounts of data.

I recommend using [DataSet.window()](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#window). See Approach 5 in [this blog post](https://mobiarch.wordpress.com/2020/11/13/preparing-time-series-data-for-rnn-in-tensorflow/) on how to do this.

**Comment by dsantiago** [comment]
The function `tf.keras.preprocessing.timeseries_dataset_from_array` is slow aswell, if you make the window in numpy, and pass the batch_size in the fit method of the model, is faster

**Comment by VincentTheBeast** [comment]
For use cases that require lining up the label with the last row rather than the first row, we can reverse the dataset using df=df.iloc[::-1] before splitting into features and labels and passing it to tf.keras.preprocessing.timeseries_dataset_from_array

The reversed row order will make not difference to the training.

**Comment by jannislanger00** [comment]
> The function `tf.keras.preprocessing.timeseries_dataset_from_array` is slow aswell, if you make the window in numpy, and pass the batch_size in the fit method of the model, is faster

yes, I was wondering why that happens. Thanks for pointing that out.
